src/realm/alloc_slab.cpp

```diff
@@ -1179,45 +1173,47 @@ void SlabAlloc::update_reader_view(size_t file_size)
                 size_t section_reservation = get_section_base(old_num_sections) - section_start_offset;
                 REALM_ASSERT(section_size == section_reservation);
                 // save the old mapping/keep it open
-                m_old_mappings.emplace_back(m_youngest_live_version, std::move(m_mappings[mapping_index]));
-                m_mappings[mapping_index] =
+                m_old_mappings.emplace_back(m_youngest_live_version, std::move(cur_entry.primary_mapping));
+                // A xover mapping cannot be present in this case:
+                REALM_ASSERT(!cur_entry.xover_mapping.is_attached());
+                cur_entry.primary_mapping =
                     util::File::Map<char>(m_file, section_start_offset, File::access_ReadOnly, section_size);
                 m_mapping_version++;
             }
 
             // 2. add any full mappings
             //  - figure out how many full mappings we need to match the requested size
             auto new_slab_base = align_size_to_section_boundary(file_size);
-            size_t num_full_mappings = get_section_index(file_size) - m_sections_in_compatibility_mapping;
-            size_t num_mappings = get_section_index(new_slab_base) - m_sections_in_compatibility_mapping;
-            size_t old_num_mappings = old_num_sections - m_sections_in_compatibility_mapping;
+            size_t num_full_mappings = get_section_index(file_size);
+            size_t num_mappings = get_section_index(new_slab_base);
+            size_t old_num_mappings = old_num_sections;
             if (num_mappings > old_num_mappings) {
                 // we can't just resize the vector since Maps do not support copy constructionn:
                 // m_mappings.resize(num_mappings);
-                std::vector<util::File::Map<char>> next_mapping(num_mappings);
+                std::vector<MapEntry> next_mapping(num_mappings);
                 for (size_t i = 0; i < old_num_mappings; ++i) {
-                    next_mapping[i] = std::move(m_mappings[i]);
+                    next_mapping[i].primary_mapping = std::move(m_mappings[i].primary_mapping);
+                    next_mapping[i].xover_mapping = std::move(m_mappings[i].xover_mapping);
+                    next_mapping[i].lowest_possible_xover_offset = m_mappings[i].lowest_possible_xover_offset;
                 }
                 m_mappings = std::move(next_mapping);
             }
 
             for (size_t k = old_num_mappings; k < num_full_mappings; ++k) {
```

Comment:
[Ignorable] Can't help wondering why `k` instead of `i` that you use just above 🤷‍♂ ?

Reply:
Diversity and inclusion. Why should 'k' be left out?